### DISCUSSION TITLE
Any advice on patching dependencies?

### DISCUSSION THREAD

**Discussion by OliverEvans96** [ipetkov/crane#192] [Q&A]
Hello,

Thanks for the super cool package.

I have a situation where an indirect dependency (mediasoup-sys) of my crate fails to build on NixOS (mainly because it tries to install its own dependencies over the network at build time). After much tinkering, I've finally found a way to patch the dependency using `buildRustPlatform`. My solution is terribly inelegant, but you can see it in my MWE [here](https://github.com/OliverEvans96/mediasoup-nix/blob/main/flake.nix#L100-L124). Hopefully there's a much easier way that I've failed to discover.

I would love to migrate from buildRustPlatform to crane for the caching gains, but I'm worried that it would be quite hard given that crane fetches dependencies outside of the normal cargo mechanisms.

Do you have any general suggestions on the subject, or specific comments on porting my solution?

Thanks!
Oliver

**Comment by dpc** [comment]
> After much tinkering, I've finally found a way to patch the dependency using `buildRustPlatform`

Did you try to engage maintainers and fix things upstream e.g. by some environment variable controlling the behavior?

**Comment by ipetkov** [comment]
Hi @OliverEvans96 thanks for the report! We don't really have a means to customize or patch individual dependency crates (especially when dealing with transitive dependencies), so the best advice I can offer at this moment is:

1. Consider submitting your patch upstream like @dpc recommended. From your example it looks fairly self-contained and hopefully the project maintainers won't have an issue accepting an environment variable to disable the automatic installation behavior
2. You could fork the crate and apply your own patch there, then update `Cargo.toml` to patch the dependency with your own fork. Though keeping up with upstream changes will require some effort, so upstreaming the change sounds like the least involved solution long term

We might be able to support a form of dependency patching in the future by copying the sources out of the nix store and into the builder working directory, but it feels like trying to put a bandaid on sandbox-unfriendly behavior which ideally should come from upstream
